src/Transform/OptimizeMemoryPools.cpp

```diff
@@ -0,0 +1,905 @@
+  topBlock->walk([&operations, &operationInLiveRange, lastLoadStore, getRef](
+                     Operation *op) {
+    // If op is a Laod/Store, of any kind, then assign it to lastLoadStore.
+    if (isLoadStoreForGetRef(getRef, op) && !operationInLiveRange)
+      operationInLiveRange = true;
+
+    if (operationInLiveRange)
+      operations.emplace_back(op);
+
+    if (op == lastLoadStore)
+      operationInLiveRange = false;
+  });
+
+  return operations;
+}
+
+/// This function returns true if `beforeOp` is visited before `op` in a
+/// traversal of the provided block.
+bool opBeforeOp(Block *block, Operation *beforeOp, Operation *afterOp) {
+  bool beforeOpIsBefore = true;
+  bool beforeOpFound = false;
+  block->walk(
+      [&beforeOpIsBefore, &beforeOpFound, beforeOp, afterOp](Operation *op) {
+        if (op == beforeOp)
+          beforeOpFound = true;
+        else if (op == afterOp && !beforeOpFound)
+          beforeOpIsBefore = false;
+      });
+  return beforeOpIsBefore;
+}
+
+/// The live range is contained between firstOp and lastOp.
+bool liveRangeIsContained(Operation *firstOp, Operation *lastOp,
+    std::vector<Operation *> liveRangeOpList) {
+  Operation *liveRangeFirstOp = liveRangeOpList[0];
+  assert(liveRangeOpList.size() > 0 &&
+         "Live range empty but must have at least one element.");
+  Operation *liveRangeLastOp = liveRangeOpList[liveRangeOpList.size() - 1];
+
+  Block *topLevelBlock = getTopBlock(firstOp);
+
+  return opBeforeOp(topLevelBlock, firstOp, liveRangeFirstOp) &&
+         opBeforeOp(topLevelBlock, liveRangeLastOp, lastOp);
+}
+
+/// Check if an operation is in the top-level block of the function.
+bool opInTopLevelBlock(Operation *op) {
+  Block *currentBlock = op->getBlock();
+
+  // If the parent operation of the current block is a FuncOp then
+  // this operation is in the top-level block.
+  return llvm::dyn_cast_or_null<FuncOp>(currentBlock->getParentOp());
+}
+
+/// Returns the outermost krnl.iterate that contains this operation.
+/// Example:
+///
+/// func() {
+///   if {
+///     krnl.iterate {  <--- Outermost loop.
+///       krnl.iterate {
+///         if {
+///           ... op ...
+///         }
+///       }
+///     }
+///   }
+/// }
+///
+Operation *getOutermostLoop(Operation *op) {
+  Operation *outermostLoop = nullptr;
+
+  // Get current block.
+  Block *currentBlock = op->getBlock();
+
+  // Current block must exist.
+  assert(currentBlock && "Operation not in a block.");
+
+  // Compute parent operation of the current block. Every block has
+  // a parent operation.
+  Operation *parentBlockOp = currentBlock->getParentOp();
+  while (!llvm::dyn_cast_or_null<FuncOp>(parentBlockOp)) {
+    if (llvm::dyn_cast_or_null<KrnlIterateOp>(parentBlockOp))
+      outermostLoop = parentBlockOp;
+    parentBlockOp = parentBlockOp->getBlock()->getParentOp();
+  }
+
+  return outermostLoop;
+}
+
+/// Returns true if two operations share the same outermost krnl.iterate.
```

Comment:
General question: if you have something like this

for(i...) {
   a = alloc(100), b = alloc(100), c = alloc(100)
   for(j) define (a)
   for(j) define(b) use(a)
   for(j) define(c) use(b)
}

would a and c be able to use the same memory? Asking this as I see you query about shared outer loops.

Reply:
They cannot share the same memory because they have a common for as their ancestor i.e. the outermost for, `for(i...)`

Reply:
> Left a couple of questions. At a general level, can you also comment on why you requires the sizes to be the same for merging? I can see how one would first try to merge all identical ones, and then progressively the ones that are most similar in sizes (e.g. 1000 would fit in 1001 bytes).

Very good point. For now I have only tried precise sizes for simplicity. This way I can compare the new incoming candidate size with any of the existing getRefs sharing that slot instead of having to keep track of the maximum MemRef size for a given slot.

Reply:
Maybe my example was too sparse:
```
for(i...) {
  a = alloc(100), b = alloc(100), c = alloc(100) 
  for(j) define (a)
  for(j) define(b) use(a)
  for(j) define(c) use(b)
  free a, b, c
}
```

Since the liveranges are confined within the I loop, there is no reason to disable this... I am thinking that the LOOP operator will introduce such things. It would be nice to have it. If you don't want to do this now, please note in the code where this constraint is expressed with a TODO.

Reply:
I see your point now. The code to handle that case is the same as the code we have today. I have temporarily disabled the optimization within sub-blocks but I can re-enable that and this case should then be covered. Slight change to the "same outermost loop" definition: the outermost loop needs to be the one in the current block rather than one at top level block i.e. the analysis should stop at the for j loops when looking to optimize this block. This will then result in a, b and c having no common outermost loop which would lead to them being able to be safely reused. The pre-condition is that the alloc and the dealloc are not hoisted outside the for i loop. If you do that, then they will be handled at the top block level.

Perhaps hoisting needs to happen after this pass.

If you hoist the allocs before this pass, the allocs would be discovered at the top level (even without enabling the sub-block level optimization). They would then be checked and pass pretty much all conditions except the outermost for loop condition. I'm not sure I see a straightforward relaxation of this condition without some heavy analysis of the loop nest.

Reply:
> If you hoist the allocs before this pass, the allocs would be discovered at the top level (even without enabling the sub-block level optimization).

I think its the opposite; as there are here, the liveranges are easily contained inside the outer loop... easier to analyze locally.

Reply:
Nope, it's not the opposite, if you hoist the alloc, it will be considered for optimization in the context that contains the outermost loop i.e. the main block of the model function.

Reply:
I've added a TODO to enable to optimization for sub-blocks.